lib/wasix/src/state/linker.rs

```diff
@@ -0,0 +1,751 @@
+// TODO: The linker *can* exist in the runtime, since technically, there's nothing that
+// prevents us from having a non-WASIX linker. However, there is currently no use-case
+// for a non-WASIX linker, so we'll refrain from making it generic for the time being.
+
+use std::{
+    collections::HashMap,
+    ffi::OsStr,
+    path::{Path, PathBuf},
+    sync::{Arc, Mutex},
+};
+
+use virtual_fs::{AsyncReadExt, FileSystem, FsError};
+use wasmer::{
+    AsStoreMut, CompileError, ExportError, Exportable, Extern, ExternType, Function, FunctionEnv,
+    FunctionEnvMut, Global, GlobalType, ImportType, Imports, Instance, InstantiationError, Memory,
+    MemoryError, Module, RuntimeError, Table, Type, Value, WASM_PAGE_SIZE,
+};
+
+use crate::{
+    fs::WasiFsRoot, import_object_for_all_wasi_versions, ModuleInitializer, WasiEnv, WasiFs,
+};
+
+use super::WasiModuleInstanceHandles;
+
+// Module handle 0 is always the main module. Side modules get handles starting from 1.
+pub static MAIN_MODULE_HANDLE: ModuleHandle = ModuleHandle(0);
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub struct ModuleHandle(u32);
+
+impl From<ModuleHandle> for u32 {
+    fn from(handle: ModuleHandle) -> Self {
+        handle.0
+    }
+}
+
+impl From<u32> for ModuleHandle {
+    fn from(handle: u32) -> Self {
+        ModuleHandle(handle)
+    }
+}
+
+const DEFAULT_RUNTIME_PATH: [&str; 3] = ["/lib", "/usr/lib", "/usr/local/lib"];
+
+pub struct MemoryAllocator {}
```

Comment:
Nit for forwards compatibility: add a `_private: ()` field so it can't be constructed manually.

Reply:
We need to add fields to this, the current 'allocate a page per module' approach is terrible. But yeah, makes sense in the meantime.